lib/v2/screens/sign_up/claim_invite/components/bottom_container.dart

```diff
@@ -0,0 +1,178 @@
+                    hintText: 'Invite code (5 words)'.i18n,
+                    hintStyle: Theme.of(context).textTheme.subtitle2OpacityEmphasis,
+                    errorText: _currentState.pageState == PageState.failure ? _currentState.errorMessage : null,
+                    suffixIcon: _inviteCodeSuffixIcon,
+                    enabledBorder: const OutlineInputBorder(
+                      borderRadius: BorderRadius.all(Radius.circular(8)),
+                      borderSide: BorderSide(
+                        width: 1,
+                        color: Colors.black,
+                      ),
+                    ),
+                    focusedBorder: const OutlineInputBorder(
+                      borderRadius: BorderRadius.all(Radius.circular(8)),
+                      borderSide: BorderSide(
+                        width: 1,
+                        color: Colors.black,
+                      ),
+                    ),
+                    focusedErrorBorder: const OutlineInputBorder(
+                      borderRadius: BorderRadius.all(Radius.circular(8)),
+                      borderSide: BorderSide(
+                        width: 1,
+                        color: AppColors.red,
+                      ),
+                    ),
+                    errorBorder: const OutlineInputBorder(
+                      borderRadius: BorderRadius.all(Radius.circular(8)),
+                      borderSide: BorderSide(
+                        width: 1,
+                        color: AppColors.red,
+                      ),
+                    ),
+                  ),
+                ),
+                const SizedBox(height: 50),
+                FlatButtonLong(
+                  title: "Claim code".i18n,
+                  onPressed: _onClaimPressed(context),
+                ),
+              ],
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+
+  Widget get _inviteCodeSuffixIcon {
+    final bool isChecked = _currentState.pageState == PageState.success && _currentState.inviteModel != null;
+    final bool canClear = !isChecked && _keyController.text.isNotEmpty;
+    final bool loading = _currentState.pageState == PageState.loading;
+
+    return QuadStateClipboardIconButton(
+      onClear: () {
+        setState(() {
+          _keyController.text = '';
+        });
+      },
+      onPaste: () async {
+        final clipboardText = await getClipboardData();
+        setState(() {
+          _keyController.text = clipboardText;
+        });
+      },
+      isChecked: isChecked,
+      canClear: canClear,
+      isLoading: loading,
+    );
+  }
+
+  void _onInviteCodeChanged() {
+    if (_keyController.text.isNotEmpty) {
+      _debouncer.run(() {
+        _signupBloc.add(ValidateInviteCode(inviteCode: _keyController.text));
+      });
+    }
+  }
+
+  VoidCallback? _onClaimPressed(BuildContext context) {
+    final bool isValid = _currentState.pageState == PageState.success && _currentState.inviteModel != null;
```

Comment:
IMO `isValid` must be a bloc state field that you can determine it's value inside the bloc. Also the snackbar can be fired from a BlocListener, `listenWhen: (_, current)=>current.isValid ==false;` try to use it for things like dialogs, snacks, navigations, initilizations values like controllers.

But they are just suggestions as does the job 👍

Reply:
I agree with this, best practices are to move all logic into the bloc and have states represent the view state. 
Not a blocker for now, but keep in mind for future reference

Reply:
Moved `isValid` into the state. I agree with handling these in the listener, however, that would be for when there is business logic and one of these actions are triggered as a result of state change. In this case, there is no business logic involved and it's purely on the UI.

Reply:
I agree with you, although, but I'd put it another way since Dart is both the UI and the logic, the question is how much logic (if-else) you want to see a file called screen (UI).